### PR TITLE
APEXCORE-276 made METRICS_TRANSPORT pluggable

### PR DESCRIPTION
**Pull request by davidyan74**
@tweise please review and merge

### CODE REVIEW

**Comment by davidyan74** [comment]
@chandnisingh please review

File: api/src/main/java/com/datatorrent/api/Context.java [hunk 1]
```diff
@@ -335,10 +335,16 @@
     Attribute<String> APPLICATION_DATA_LINK = new Attribute<String>(new String2String());
     /**
      * Transport to push the stats and the metrics, "builtin:{topic}" if STRAM should push the data directly
-     * using websocket with the given topic
+     * using websocket with the given topic.
+     * If using a custom transport, please use Object2String codec to specify the transport object.  The object must
+     * be from a class that implements the AutoMetric.Transport interface
      */
     Attribute<String> METRICS_TRANSPORT = new Attribute<String>(new String2String());
```
Comment (otterc) [inline api/src/main/java/com/datatorrent/api/Context.java]: In the javadoc for the attribute Object2String codec is mentioned but in the declaration it is String2String

**Comment by otterc** [comment]
A recommendation is to add a test which sets a custom metrics transport and see if that gets picked up

File: api/src/main/java/com/datatorrent/api/Context.java (hunk 1 above)
Comment (davidyan74) [inline api/src/main/java/com/datatorrent/api/Context.java]: @chandnisingh this is for backward compatibility for the value "builtin:{topic}"

Comment (otterc) [inline api/src/main/java/com/datatorrent/api/Context.java]: So what it means here is that we are asking the user to convert the object to string using Object2String codec before setting this attribute?
Why don't we deprecate this and add another attribute which is of type `Attribute<AutoMetric.Transport>` which is consistent with  rest of the complex attributes?

**Comment by tweise** [comment]
This is why is is important to not assume things as stable right of the bat. Is this attribute in its present form used anywhere? I don't think an end user would even be able to understand what to do with it? If so, we should consider it a bug and fix it.

**Comment by davidyan74** [comment]
test is added

File: api/src/main/java/com/datatorrent/api/Context.java (hunk 1 above)
Comment (davidyan74) [inline api/src/main/java/com/datatorrent/api/Context.java]: I originally wanted to do this, but how do we support this while supporting the built-in websocket transport at the same time? Looking for suggestions on how to do this a better way.

File: api/src/main/java/com/datatorrent/api/Context.java [hunk 2]
```diff
@@ -335,10 +335,16 @@
...
     /**
+     * The interval in milliseconds to resend metrics schema. Default value is 10,000 (10 seconds)
+     */
+    Attribute<Long> METRICS_SCHEMA_RESEND_INTERVAL = new Attribute<Long>(10000L);
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/Context.java]: Shouldn't the resend interval be part of the transport as the need to do that is driven by the properties of the transport, i.e. need to resend if messages could be lost?

File: api/src/main/java/com/datatorrent/api/Context.java (hunk 1 above)
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/Context.java]: As discussed, maybe look at the other option again to use a placeholder object for the "built-in" transport and swap it out for the engine internal implementation when wiring the dependencies in the AM?

**Comment by davidyan74** [comment]
I just implemented what Thomas suggested, but I kept the METRICS_TRANSPORT attribute since it's not currently widely used but I can change it if people have concerns.

File: engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java [hunk 3]
```diff
@@ -970,4 +969,48 @@ public void onClose(int closeCode, String message)
...
+    StreamingContainerManager dnmgr = lc.dnmgr;
+    StramAppContext appContext = new StramTestSupport.TestAppContext();
+
+    AppDataPushAgent pushAgent = new AppDataPushAgent(dnmgr, appContext);
+    pushAgent.init();
+    pushAgent.pushData();
+    Assert.assertTrue(TestMetricTransport.messages.size() > 0);
+    pushAgent.close();
+    String msg = TestMetricTransport.messages.get(0);
+    System.out.println("Got this message: " + msg);
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java]: Not needed.

**Comment by davidyan74** [comment]
Addressed all PR comments and rebased.

**Comment by otterc** [comment]
There is no change made to japicmp config to exclude the attribute from semver, so just was wondering why it didn't complain. I don't have any problem with this change just curious to know if japicmp doesn't check for parameterized types

**Pull request merged**